onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc

```diff
@@ -0,0 +1,356 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "nnapi_execution_provider.h"
+
+#include "builders/model_builder.h"
+#include "core/framework/allocatormgr.h"
+#include "core/framework/compute_capability.h"
+#include "core/graph/model.h"
+#include "core/session/onnxruntime_cxx_api.h"
+#include "core/session/inference_session.h"
+
+namespace onnxruntime {
+
+constexpr const char* NNAPI = "Nnapi";
+
+NnapiExecutionProvider::NnapiExecutionProvider()
+    : IExecutionProvider{onnxruntime::kNnapiExecutionProvider} {
+  DeviceAllocatorRegistrationInfo device_info{
+      OrtMemTypeDefault,
+      [](int) { return onnxruntime::make_unique<CPUAllocator>(
+                    onnxruntime::make_unique<OrtMemoryInfo>(
+                        NNAPI, OrtAllocatorType::OrtDeviceAllocator)); },
+      std::numeric_limits<size_t>::max()};
+
+  InsertAllocator(CreateAllocator(device_info));
+
+  DeviceAllocatorRegistrationInfo cpu_memory_info({OrtMemTypeCPUOutput,
+                                                   [](int) { return onnxruntime::make_unique<CPUAllocator>(onnxruntime::make_unique<OrtMemoryInfo>(NNAPI, OrtAllocatorType::OrtDeviceAllocator, OrtDevice(), 0, OrtMemTypeCPUOutput)); },
+                                                   std::numeric_limits<size_t>::max()});
+
+  InsertAllocator(CreateAllocator(cpu_memory_info));
+}
+
+NnapiExecutionProvider::~NnapiExecutionProvider() {}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+NnapiExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
+                                      const std::vector<const KernelRegistry*>& /*kernel_registries*/) const {
+  std::vector<std::unique_ptr<ComputeCapability>> result;
+  // Handle If and Loop operators
+  if (graph.IsSubgraph()) {
+    return result;
+  }
+
```

Comment:
Why can't this EP take nodes in a subgraph? They're executed the same way as the main graph so it should be fine to look for nodes that can be handled by NNAPI.

Reply:
Added comments, this will need more investigation, we do not support it for now

Reply:
FWIW it should work just fine. Main potential issue I can think of would be if it needs initializers from the parent graph, but those can be read with graph_utils::GetConstantInitializer once you switch to using Graph.

---
In reply to: [444567659](https://github.com/microsoft/onnxruntime/pull/4287#discussion_r444567659) [](ancestors = 444567659)